python/swiflow/_common.py

```diff
@@ -268,9 +292,100 @@ def decode_err(self, err: ValueError) -> ValueError:
         except ValueError as e:
             raise self.decode_err(e) from None
+
+
+def _infer_layers_impl(gd: nx.DiGraph[_V]) -> Mapping[_V, int]:
+    """Fix flow layers one by one depending on order constraints."""
+    pred = {u: set(gd.predecessors(u)) for u in gd.nodes}
+    work = {u for u, pu in pred.items() if not pu}
+    ret: dict[_V, int] = {}
+    for l_now in itertools.count():
+        if not work:
+            break
+        next_work: set[_V] = set()
+        for u in work:
+            ret[u] = l_now
+            for v in gd.successors(u):
+                ent = pred[v]
+                ent.discard(u)
+                if not ent:
+                    next_work.add(v)
+        work = next_work
+    if len(ret) != len(gd):
+        msg = "Failed to determine layer for all nodes."
+        raise ValueError(msg)
+    return ret
+
+
+def _is_special(
+    pp: PPlane | None,
+    in_fu: bool,  # noqa: FBT001
+    in_fu_odd: bool,  # noqa: FBT001
+) -> bool:
+    if pp == PPlane.X:
+        return in_fu
+    if pp == PPlane.Y:
+        return in_fu and in_fu_odd
+    if pp == PPlane.Z:
+        return in_fu_odd
+    return False
+
+
+def _special_edges(
+    g: nx.Graph[_V],
+    anyflow: Mapping[_V, _V | AbstractSet[_V]],
+    pplane: Mapping[_V, PPlane] | None,
+) -> set[tuple[_V, _V]]:
+    """Compute special edges that can bypass partial order constraints in Pauli flow."""
+    ret: set[tuple[_V, _V]] = set()
+    if pplane is None:
+        return ret
+    for u, fu_ in anyflow.items():
+        fu = fu_ if isinstance(fu_, AbstractSet) else {fu_}
+        fu_odd = odd_neighbors(g, fu)
+        for v in itertools.chain(fu, fu_odd):
+            if u == v:
+                continue
+            if _is_special(pplane.get(v), v in fu, v in fu_odd):
+                ret.add((u, v))
+    return ret
+
+
+def infer_layers(
+    g: nx.Graph[_V],
+    anyflow: Mapping[_V, _V | AbstractSet[_V]],
+    pplane: Mapping[_V, PPlane] | None = None,
+) -> Mapping[_V, int]:
+    """Infer layer from flow/gflow using greedy algorithm.
+
+    Parameters
+    ----------
+    g : `networkx.Graph`
+        Simple graph representing MBQC pattern.
+    anyflow : `tuple` of flow-like/layer
+        Flow to verify. Compatible with both flow and generalized flow.
+    pplane : `collections.abc.Mapping`, optional
+        Measurement plane or Pauli index.
+
+    Notes
+    -----
+    This function operates in Pauli flow mode only when :py:obj`pplane` is explicitly given.
+    """
+    gd: nx.DiGraph[_V] = nx.DiGraph()
+    gd.add_nodes_from(g.nodes)
+    special = _special_edges(g, anyflow, pplane)
+    for u, fu_ in anyflow.items():
+        fu = fu_ if isinstance(fu_, AbstractSet) else {fu_}
+        fu_odd = odd_neighbors(g, fu)
+        for v in itertools.chain(fu, fu_odd):
+            if u == v or (u, v) in special:
+                continue
+            gd.add_edge(u, v)
+    gd = gd.reverse()
+    return _infer_layers_impl(gd)
```

Comment:
It is 2.3× faster to directly compute the mappings `pred` and `succ`, without computing the intermediate graph.
```suggestion
    pred: dict[_V, set[_V]] = {v: set() for v in g.nodes}
    succ: dict[_V, set[_V]] = {v: set() for v in g.nodes}
    special = _special_edges(g, anyflow, pplane)
    for u, fu in anyflow.items():
        fu_set = fu if isinstance(fu, AbstractSet) else {fu}
        fu_odd = odd_neighbors(g, fu_set)
        for v in (*fu_set, *fu_odd):
            if u != v and (u, v) not in special:
                pred[u].add(v)
                succ[v].add(u)
    work = {u for u, pu in pred.items() if not pu}
    ret: dict[_V, int] = {}
    for l_now in itertools.count():
        if not work:
            break
        next_work: set[_V] = set()
        for u in work:
            ret[u] = l_now
            for v in succ[u]:
                ent = pred[v]
                ent.discard(u)
                if not ent:
                    next_work.add(v)
        work = next_work
    if len(ret) != len(g.nodes):
        msg = "Failed to determine layer for all nodes."
        raise ValueError(msg)
    return ret
```

Reply:
Merged in ee90522f4560c7d0c75277f2d03a8b4b89953b63 .